sklearn/model_selection/_search.py

```diff
@@ -320,24 +323,50 @@ def fit_grid_point(X, y, estimator, parameters, train, test, scorer,
     return score, parameters, n_samples_test
 
 
-def _check_param_grid(param_grid):
-    if hasattr(param_grid, 'items'):
-        param_grid = [param_grid]
+def _check_param_grid_or_dist(param_grid_or_dist):
+    """Validate param_grid/distribution and return the unique parameters"""
+    parameter_names = set()
 
-    for p in param_grid:
+    if hasattr(param_grid_or_dist, 'items'):
+        param_grid_or_dist = [param_grid_or_dist]
+
+    for p in param_grid_or_dist:
         for v in p.values():
+            if not isinstance(v, (list, tuple, np.ndarray, rv_continuous,
+                                  rv_discrete, rv_frozen)):
+                raise ValueError("Parameter values should be an array-like or "
+                                 "one of scipy.distributions."
+                                 "rv{continuous|discrete|frozen}. %r is "
+                                 "not one of those." % v)
+
             if isinstance(v, np.ndarray) and v.ndim > 1:
                 raise ValueError("Parameter array should be one-dimensional.")
 
-            check = [isinstance(v, k) for k in (list, tuple, np.ndarray)]
-            if True not in check:
-                raise ValueError("Parameter values should be a list.")
-
-            if len(v) == 0:
+            if isinstance(v, (list, tuple, np.ndarray)) and len(v) == 0:
                 raise ValueError("Parameter values should be a non-empty "
                                  "list.")
 
+        parameter_names.update(p.keys())
+
+    return list(parameter_names)
+
+
+def _get_metric_name(scoring):
+    """Generate the metric name given the scoring parameter"""
+    if callable(scoring):
+        if scoring.__name__ == "_passthrough_scorer":
+            return "estimator_default_scorer"
+        else:
+            return "custom_metric_%s" % (scoring.__name__,)
```

Comment:
In the multiple metrics context, there can still be naming conflicts. In any case, using `__name__` here is a bit of a hack, and most custom scorers will be `_PredictScorer` or similar.